app/services/foreman_discovery/fact_to_category_resolver.rb

```diff
@@ -0,0 +1,98 @@
+module ForemanDiscovery
+  class FactToCategoryResolver
+    include ActionView::Helpers::NumberHelper
+
+    attr_reader :categories, :interfaces
+
+    CATEGORIES_NAMES = [N_("Highlights"), N_("Storage"), N_("Hardware"), N_("Network"), N_("Software"), N_("IPMI"), N_("Miscellaneous")].freeze
+
+    def initialize(host)
+      categories = %i[highlights storage hardware network software ipmi]
+
+      @regex_array = categories.map do |category|
+        settings_category = settings_discovery_fact_prefix(category)
+        if settings_category.empty?
+          send(category)
+        else
+          Regexp.new(settings_category)
+        end
+      end
+
+      @regex_array << false
+
+      @categories = Array.new(7) { {} }
```

Comment:
Can you remove this preliminary optimization? I know you copied the code, it's not good to do this.